ciao-launcher/qemu.go

```diff
@@ -289,57 +289,76 @@ func (q *qemu) deleteImage() error {
 	return nil
 }
 
-func computeTapParam(vnicName string, networkNode bool) (string, *os.File, error) {
-	if !networkNode {
-		return fmt.Sprintf("tap,ifname=%s,script=no,downscript=no",
-			vnicName), nil, nil
-	}
+func computeMacvtapParam(vnicName string, mac string, queues int) ([]string, []*os.File, error) {
+
+	fds := make([]*os.File, queues)
+	params := make([]string, 0, 8)
 
 	ifIndexPath := path.Join("/sys/class/net", vnicName, "ifindex")
 	fip, err := os.Open(ifIndexPath)
 	if err != nil {
 		glog.Errorf("Failed to determine tap ifname: %s", err)
-		return "", nil, err
+		return nil, nil, err
 	}
 	defer fip.Close()
 
 	scan := bufio.NewScanner(fip)
 	if !scan.Scan() {
 		glog.Error("Unable to read tap index")
-		return "", nil, fmt.Errorf("Unable to read tap index")
+		return nil, nil, fmt.Errorf("Unable to read tap index")
 	}
 
 	i, err := strconv.Atoi(scan.Text())
 	if err != nil {
 		glog.Errorf("Failed to determine tap ifname: %s", err)
-		return "", nil, err
+		return nil, nil, err
 	}
 
-	tapDev := fmt.Sprintf("/dev/tap%d", i)
+	//mq support
+	fdParam := ""
+	fdSeperator := ""
+	for q := 0; q < queues; q++ {
 
-	f, err := os.OpenFile(tapDev, os.O_RDWR, 0666)
-	if err != nil {
-		glog.Errorf("Failed to open tap device %s: %s", tapDev, err)
-		return "", nil, err
-	}
+		tapDev := fmt.Sprintf("/dev/tap%d", i)
 
-	/*
-	   3, what do you mean 3.  Well, it turns out that files passed to child
-	   processes via cmd.ExtraFiles have different fds in the child and the
-	   parent.  In the child the fds are determined by the file's position
-	   in the ExtraFiles array + 3. Since we're only specifying a single
-	   file we end up with an child fd of 3.
-	*/
+		f, err := os.OpenFile(tapDev, os.O_RDWR, 0666)
+		if err != nil {
+			glog.Errorf("Failed to open tap device %s: %s", tapDev, err)
```

Comment:
Won't we leak fds here if we get an error, i.e, if we fail to open a file for the 3rd queue, but the first 2 succeed.